### DISCUSSION TITLE
How to use JUMP ?

### DISCUSSION THREAD

**Discussion by volga629-1** [FoobarOy/foomuuri#31] [General]
Example 

rule like this 

```
 {"rule": {"family": "ip", "table": "filter", "chain": "rtpengine", "handle": 33, "expr": [{"xt": {"type": "target", "name": "RTPENGINE"}},
```

```
macro {

 rtp-udp       udp dport 1000-2000;
}

chain public-localhost {
          rtp-udp jump rtp
}

chain rtp {
      {"rule": {"family": "ip", "table": "filter", "chain": "rtp", "handle": 33, "expr": [{"xt": {"type": "target", "name": "RTPENGINE"}},

{

**Comment by kimheino** [comment]
Create `/etc/foomuuri/rtp.nft` file to define rtp-chain:

```
chain rtp {
  ...your raw nft rules here, or empty if created dynamically...
}
```

Then jump to it in `foomuuri.conf`:

```
macro {
  rtp-udp       udp dport 1000-2000
}

chain public-localhost {
  rtp-udp nft "jump rtp"
}
```

Above example will jump to rtp for both IPv4 and IPv6 traffic. Use this if you want to jump for IPv4 only:

```
  ipv4 rtp-udp nft "jump rtp"
```

**Comment by volga629-1** [comment]
Thank you for explanation. I am discovering all possibilities.

**Comment by volga629-1** [comment]
I am trying rule like this 

```
# From wan to fw
wan1-fw {
        ipv4 sip-ext-udp saddr @ipv4_CA nft "jump ips"
        ipv4 sip-ext-udp saddr @ipv4_US nft "jump ips"
        # global deny rule
        reject log
}
```

That produce error 


```
foomuuri check
/var/lib/foomuuri/next.fw:143:43-45: Error: Could not process rule: Operation not supported
		ip saddr @ipv4_CA_4 udp dport 5088 jump ips
		                                        ^^^
```

**Comment by kimheino** [comment]
Most probably your `ips` chain is not defined correctly. It must be rule-only chain, without  `type ... hook` line. For example:

```
chain ips {
   tcp dport 22 accept
   tcp dport 80 accept
}
```

**Comment by volga629-1** [comment]
Yes , right on, thank you 

why this line is important ?

```
#type filter hook forward priority filter + 10; policy accept;

```

Is firewall make decision ?

Custom chain 

```
chain ips {
        #type filter hook forward priority filter + 10; policy accept;
        log prefix "Inspect SIP2IPS: " group 3
        counter packets 0 bytes 0
        queue flags fanout to 4-8
}

```

**Comment by kimheino** [comment]
> why this line is important ?
> #type filter hook forward priority filter + 10; policy accept;

Only "top level" chains in nftables can and must use `hook`, not
sub-chains that you jump to. Foomuuri defines it and then jumps to its
own and your custom sub-chains. Simply put, just leave it out and it
works.

For details see for example:

https://thermalcircle.de/doku.php?id=blog:linux:nftables_packet_flow_netfilter_hooks_detail

**Comment by volga629-1** [comment]
Thank you for explanation.

**Comment by kimheino** [comment]
Next version will support "queue" natively. Your above "jump + chain ips" can be written as:

```
forward {
  log "Inspect SIP2IPS" log_level "group 3"
  counter
  queue flags fanout to 4-8
}
```
